### ISSUE TITLE
Can we get more schedulers for flow based models such as SD3, SD3.5, and flux

### ISSUE THREAD

**Issue by linjiapro** [huggingface/diffusers#9924]
It seems advanced schedulers such as DDIM, and the dpm++ 2m does work with flow based model such as SD3, SD3.5, and flux. 
However, I only see 2 flow based schedulers in diffusers codebase:

FlowMatchEulerDiscreteScheduler, and'
FlowMatchHeunDiscreteScheduler

I tried to use DPMSolverMultistepScheduler, but it does not generate correct images with flow based models. Help?

**Comment by sayakpaul** [comment]
Cc: @asomoza @hlky @yiyixuxu

**Comment by ukaprch** [comment]
I'm glad this shortcoming has been brought to light. With that in mind, I have been testing a new diffuser scheduler which is adapted from the current diffuser / schedulers library (scheduling_dpmsolver_multistep.py) along with this: [https://github.com/crowsonkb/k-diffusion/blob/master/k_diffusion/sampling.py](url)
the net result is a new scheduler containing 7 different samplers ("dpmsolver2", "dpmsolver2A", "dpmsolver++2M", "dpmsolver++2S", "dpmsolver++sde", "dpmsolver++2Msde", "dpmsolver++3Msde).
This new scheduler would work for both the existing FLUX and SD3 pipelines in the diffusers library although I will say that SD3 suffers (temperamental) somewhat when using karras, exponential and lambdas sigmas at times.
I have done a lot of empirical testing with it and it appears to function as intended. You all are most welcome to try it.
I have attached the scheduler and the parameters I use. Some of the new parameters will become apparent to you as you familiarize yourself with the scheduler's use. As things stand now, SD3 benefits somewhat from using the SDXL approach in creating the sigmas vs the new FLUX approach. See the scheduler for details on that. I found that SD3 can also run with dynamic shifting as FLUX does, but again results can vary.

Create the pipeline as you normally would then append the scheduler with the required parameters:
Note: you can also modify parameters such as s_noise ,use_noise_sampler, etc.
[scheduling_flow_match_dpmsolver_multistep.txt](https://github.com/user-attachments/files/17777948/scheduling_flow_match_dpmsolver_multistep.txt)


SD3:
            pipe.scheduler = FlowMatchDPMSolverMultistepScheduler.from_config(pipe.scheduler.config,solver_order=2,sigma_schedule='karras',algorithm_type="dpmsolver2")

FLUX:
            pipe.scheduler = FlowMatchDPMSolverMultistepScheduler.from_config(pipe.scheduler.config,solver_order=2,sigma_schedule='karras',algorithm_type="dpmsolver2")

**Comment by linjiapro** [comment]
@ukaprch 

From the import clauses txt file, it seems the classes are already under diffusers repo for the testing runs. It seems to be suitable to include the file to the official diffusers repo?

**Comment by ukaprch** [comment]
Some of you can update your diffusers library to incorporate the scheduler I proposed (which is a first start) for independent testing (as I have). The scheduler as written can be incorporated into the diffuser library w/o change and will work with the current FLUX and SD3 pipelines.

**Comment by linjiapro** [comment]
@sayakpaul @ukaprch 

I have a pending PR which has not been merged: 
https://github.com/huggingface/diffusers/pull/9758

Once above merges, I can work on sending another PR related to schedulers. In the meanwhile, anyone else are welcomed to raise a pull request.

**Comment by vladmandic** [comment]
isn't this a duplicate of #9607 ?

**Comment by vladmandic** [comment]
@ukaprch why not create a pr to include this in diffusers?
i've integrated your code in sdnext, dpm flowmatch is working fine except any kind of karras/exponential results in full blur.
anyhow, here are some sample grids for sd35 and flux.1

![dpm-flowmatch-sd35](https://github.com/user-attachments/assets/05ecd035-43d1-49e0-9a11-8cab6181d3ee)
![dpm-flowmatch-flux](https://github.com/user-attachments/assets/d5017825-15bc-41d3-a567-053b9692b24a)

**Comment by ukaprch** [comment]
Thanks for testing and your feedback. Please note that I had no such problems using the FLUX pipeline with the new scheduler using algorithms (karras, exponential, lambdas) see attached.

            image = pipe(
                prompt='a cat holding a sign that says hello world',
                width = 1024,
                height = 1024,
                guidance_scale=3.5,
                generator=generator, (seed = 114747598)
                max_sequence_length=512,
                num_inference_steps=30).images[0]

I ran the above pipeline using the basic FLUX dev prompt from Replicate and generated the following:
![image](https://github.com/user-attachments/assets/e4c0a0ff-37e2-4c34-a4d1-25ab296e7300)

**Comment by vladmandic** [comment]
bit more testing, karras and exponential are working fine for flux, but not for sd35

flux:
![image](https://github.com/user-attachments/assets/5403ccfa-a033-4e38-977b-f5b9fd1ccdaa)

sd35
![image (1)](https://github.com/user-attachments/assets/2bd85010-d4f3-4ffe-a569-b9ceedd205d2)

actual params used:
```log
10:50:38-967529 DEBUG    Sampler: sampler="DPM2++ 2M FlowMatch" class="FlowMatchDPMSolverMultistepScheduler config={'num_train_timesteps': 1000, 'shift': 1, 'use_dynamic_shifting': False, 'solver_order': 2, 'sigma_schedule': 'karras', 'use_SD35_sigmas': True, 'algorithm_type': 'dpmsolver++2M', 'use_noise_sampler': True}
```

**Comment by ukaprch** [comment]
OK, I noticed you are using a shift: 1. This should be shift: 3 which is the std for SD3. I'll look into some more.

**Comment by vladmandic** [comment]
> I found with FLUX that you need generally 25 or more steps to get good images.

actually, the worst-case scenario here is sd35-medium ;)
flux is behaving better in this regard.

re: shift - i've noticed. i have it as user-configurable item

> So what are you thinking? Are we almost good to proceed with this as a scheduler?

definitely!

from a quick look at the code, 99% is clean.
`del sigmas1` looks unsafe as it may not be defined in all code-paths. and in general, does it help at all to have all those del statements in `set_timestep()`?

i wish if `use_SD35_sigmas` could be auto-determined, but i don't see a clean way.
but it does need a better name, definitely not something with mixed-capitalization :)

**Comment by ukaprch** [comment]
vladmandic: As noted, this schedule is not designed for flow match derived sampling.

**Comment by linjiapro** [comment]
@ukaprch 

Let me know if it is OK that I send out a PR includes the script you offered, if you can do it, I will not send out the PR.

**Comment by vladmandic** [comment]
@ukaprch i agree with all of your comments, but i feel like something went wrong in the last version?
comparing code from last week to current one:
![old](https://github.com/user-attachments/assets/474b7ab8-7aae-4890-bbe7-0ce40db071b5)
![new](https://github.com/user-attachments/assets/77669297-7c1f-4744-a3bd-2aab824bf00f)

generated using sd35-medium and example config i'm using:
```json
{'num_train_timesteps': 1000, 'beta_start': 0.0001, 'beta_end': 0.02, 'beta_schedule': 'linear', 'shift': 3, 'use_dynamic_shifting': False, 'solver_order': 2, 'sigma_schedule': 'lambdas', 'use_beta_sigmas': True, 'algorithm_type': 'dpmsolver2', 'use_noise_sampler': True}
```

**Comment by vladmandic** [comment]
also, with flux and sigma_method=betas, i'm getting index-out-of-range at 
> sigma_next = self.sigmas[self.step_index + 1]

**Comment by ukaprch** [comment]
> also, with flux and sigma_method=betas, i'm getting index-out-of-range at

sigma_next = self.sigmas[self.step_index + 1]

I was unable to replicate your error using this config with the scheduling_flow_match_dpmsolver_multistep.py
this condition should never occur b/c we append a zero sigma at the end "in all cases" save for use of "sigma_min" in the original scheduler. my config:

◢ | config | OrderedDict([('num_train_timesteps', 1000), ('beta_start', 0.0001), ('beta_end', 0.02), ('beta_schedule', 'linear'), ('trained_betas', None), ('solver_order', 2), ('algorithm_type', 'dpmsolver2'), ('solver_type', 'midpoint'), ('sigma_schedule', 'betas'), ('shift', 3.0), ('midpoint_ratio', 0.5), ('s_noise', 1), ('use_noise_sampler', True), ('use_beta_sigmas', True), ('use_dynamic_shifting', True), ('base_shift', 0.5), ('max_shift', 1.15), ('base_image_seq_len', 256), ('max_image_seq_len', 4096), ('_use_default_values', ['trained_betas', 'solver_type', 'solver_order'])]) | FrozenDict

Also just so we're on the same page:

Maybe some confusion on this but sigma_schedule = "betas" is not the same thing as
use_beta_sigmas=True.

self.config.sigma_schedule == "betas"
use_beta_sigmas	True
+		sigmas	tensor([1.0000, 0.9979, 0.9932, 0.9865, 0.9779, 0.9676, 0.9554, 0.9413, 0.9253,
        0.9072, 0.8869, 0.8642, 0.8389, 0.8109, 0.7799, 0.7456, 0.7079, 0.6665,
        0.6211, 0.5718, 0.5184, 0.4611, 0.4003, 0.3368, 0.2718, 0.2074, 0.1462,
        0.0923, 0.0509, 0.0309, 0.0000], device='cuda:0', dtype=torch.float64)	Tensor

self.config.sigma_schedule == None
use_beta_sigmas	True
+		sigmas	tensor([1.0000, 0.9889, 0.9773, 0.9651, 0.9523, 0.9388, 0.9246, 0.9096, 0.8937,
        0.8769, 0.8590, 0.8401, 0.8199, 0.7983, 0.7753, 0.7506, 0.7241, 0.6955,
        0.6646, 0.6311, 0.5947, 0.5550, 0.5115, 0.4635, 0.4106, 0.3516, 0.2857,
        0.2115, 0.1273, 0.0309, 0.0000], device='cuda:0', dtype=torch.float64)	Tensor

What did I miss?

**Comment by vladmandic** [comment]
> Maybe some confusion on this but sigma_schedule = "betas" is not the same thing as use_beta_sigmas=True.

i know - use_beta_sigmas is to make sd35 happy. sigma_schedule=betas is new sigma method that you didn't have before.

**Comment by ukaprch** [comment]
Yes. The degradation problem is most probably due to the way FlowMatch approaches the problem. The sigmas / timesteps are vastly compressed under FlowMatch which negates their usability. More steps alleviates the problem at a cost. This cannot be stressed enough for new / existing users that use these type schedulers. I'm very happy with most of the results I've achieved using them. As for SD3.5 Large, it more resembles SDXL than it does FLUX. Not sure why they chose (3) text encoders. Without the beta sigmas it would be far worse. I've also played around with using dynamic shifting with SD3.5 large. It requires you to add the required functionality to the pipeline to make it work. Something to think about I guess.

**Comment by linjiapro** [comment]
@ukaprch, 

I believe you have studied this extensively.

Can we add some best usage pattern to the code, they can be comments on top of those schedulers, and the samples can be:

if you use this for SD35, here is the recommended way:

......

**Comment by linjiapro** [comment]
After the most recommended usage pattern is added, I plan to send out a PR.

**Comment by vladmandic** [comment]
@ukaprch i agree with your thoughts on sd35, especially on their choice of both clip-l and clip-g for no reason.
but what i was referring to as "degradation" is that with sd35 your latest code performs worse than your earlier code - that is especially visible with karras.
old vs new
![image](https://github.com/user-attachments/assets/d80aed83-6d86-4fc8-8904-5d80da47fa5f)
![image](https://github.com/user-attachments/assets/d261fef9-f6d4-4d7a-94ac-f92d0ba19432)

**Comment by ukaprch** [comment]
Can you give me all the parameters you used for the above images so I can see?

**Comment by vladmandic** [comment]
> ```json
> {'num_train_timesteps': 1000, 'beta_start': 0.0001, 'beta_end': 0.02, 'beta_schedule': 'linear', 'shift': 3, 'use_dynamic_shifting': False, 'solver_order': 2, 'sigma_schedule': 'karras', 'use_beta_sigmas': True, 'algorithm_type': 'dpmsolver2', 'use_noise_sampler': True}
> ```

**Comment by ukaprch** [comment]
I needed to setup my environment for sd35 Medium which I never tested. Having done that, **using your same parameters** I ran both sd35 Medium & Large for 30 and 40 steps respectively and encountered no problems:

![SD35 MEDIUM](https://github.com/user-attachments/assets/cb73982f-ae71-47f0-a9a3-a7b0457347d5)
![SD35 LARGE](https://github.com/user-attachments/assets/058fee5f-0aeb-4540-a0cd-62f3f9de53aa)

Did something change in your environment when you updated? Did (perhaps) any code get misaligned in upgrading and not running the same?

As an aside, those using sd 35 should be aware that these models will not run well using FLUX aspect ratios and sizes. Case in point, I ran sd 35 Large using aspect ratio 2:3  with size: 1152 x 1728. As you can see in the image below both the top and bottom of the image contain artifacts reflecting problems in generating this size image. I also ran into the same problem if I ran an image with aspect ratio: 1:1 size: 1408 X 1408 which FLUX can easily handle.

![SD35 2-3 1152 x 1728 aspect ratio problem ](https://github.com/user-attachments/assets/51a3941e-05e0-4bb5-8857-3f1818377e72)

**Comment by vladmandic** [comment]
ah, my bad, the reason was that your older implementation had hard-coded beta start/end values while newer had them as params and my implementation auto-detected that param was available and it was setting it to user-value which is not the same.
when beta start/end are set to same values, both versions act identically.

**Comment by vladmandic** [comment]
fyi, i've fully integrated your code in sdnext (with credits)

**Comment by ukaprch** [comment]
I pulled a fast one on you. :-) New approach with variable betas makes more sense. I myself prefer the Custom Betas: start=0.00085; end=0.012 for most of my images, but the default is good to. I also was able to produce a reasonable karras beta image in 30 steps. More steps == more contrast it seems.
@linjiapro @vladmandic Sooooo, are we more or less good to go?

**Comment by yiyixuxu** [comment]
@ukaprch @linjiapro @vladmandic 
thanks for the discussion! Welcome a PR if any of you have time & interested!

**Comment by vladmandic** [comment]
@ukaprch re: scripy - yup, i saw it, not an issue. overall, yes, i'd say we're good-to-go!

**Comment by linjiapro** [comment]
Hi All, I am happy to send out a PR. I might do that next week given this week is US thanksgiving. Thanks. If anyone is interested in doing this earlier, go ahead!

**Comment by yiyixuxu** [comment]
the awesome @lawrence-cj is adding dpm flow match from this PR if any of you want to check it out https://github.com/huggingface/diffusers/pull/9982

cc @vladmandic here

**Comment by ukaprch** [comment]
@yiyixuxu  @vladmandic @linjiapro
I have good news coming. I may have a much improved version of this scheduler and the FlowMatchEulerDiscrete scheduler. I changed a few things, slightly new approach and the results are very promising. Stay tuned.

**Comment by linjiapro** [comment]
It seems the PR yiyixuxu mentioned added a flow match multi step solver. We could combine wisdom here to add the outcomes of this discussion there to even further enhance that PR :)

**Comment by yiyixuxu** [comment]
look forward to your PR! @ukaprch

**Comment by ukaprch** [comment]
comment to keep post alive!!!!!!!!!!!!!

**Comment by a-r-r-o-w** [comment]
Added a `wip` label so that this does not get marked as stale any more. Gentle ping to @hlky @yiyixuxu
